### PR TITLE
discussion feed-back:

### PR DESCRIPTION
**Pull request by dkaukov**
1) Merged factory beans to one
2) fixed generics flow
3) LoggingUtils private.
4) EnvironmentProvider is passed to the interceptors.
5) Configurable sampler

### CODE REVIEW

File: src/main/java/com/opentable/kafka/builders/KafkaBaseBuilder.java [hunk 1]
```diff
@@ -148,6 +150,10 @@ void withLogging(boolean enabled) {
     void withSamplingRatePer10Seconds(final int rate) {
         loggingSampleRate = rate;
     }
+    void withSamplingRate(final int rate) {
```
Comment (mikebell90) [inline src/main/java/com/opentable/kafka/builders/KafkaBaseBuilder.java]: This is a bit confusing. Two similar methods, etc. Rename to withRandomSamplingRate. Also the withSamplingRate shhould reverse the type back.

File: src/main/java/com/opentable/kafka/builders/KafkaBuilderConfiguration.java [hunk 2]
```diff
@@ -25,8 +25,7 @@
  */
 @Configuration
 @Import({
-        KafkaConsumerBuilderFactoryBean.class,
-        KafkaProducerBuilderFactoryBean.class,
+        KafkaBuilderFactoryBean.class,
```
Comment (mikebell90) [inline src/main/java/com/opentable/kafka/builders/KafkaBuilderConfiguration.java]: The annotation now must be renamed. It's singular, not plural now.

File: src/main/java/com/opentable/kafka/builders/KafkaBuilderFactoryBean.java [hunk 3]
```diff
@@ -67,4 +71,36 @@
         return originalMap;
     }
 
+
+    public KafkaConsumerBuilder<?, ?> consumerBuilder() {
```
Comment (mikebell90) [inline src/main/java/com/opentable/kafka/builders/KafkaBuilderFactoryBean.java]: Why are these not K,V. Please fix. This worked before.

File: src/main/java/com/opentable/kafka/builders/KafkaConsumerBuilder.java [hunk 4]
```diff
@@ -54,10 +54,10 @@
     private Class<? extends PartitionAssignor> partitionStrategy = RangeAssignor.class;
     // Kafka is really stupid. In the properties you can only configure a no-args
     // and then they hack around it if you have one supplied
-    private Class<? extends Deserializer<K>> keyDe;
-    private Class<? extends Deserializer<V>> valueDe;
-    private Deserializer<K> keyDeserializerInstance;
-    private Deserializer<V> valueDeserializerInstance;
+    private Class<? extends Deserializer> keyDe;
```
Comment (mikebell90) [inline src/main/java/com/opentable/kafka/builders/KafkaConsumerBuilder.java]: Don't remove the K,V. There's no need.

File: src/main/java/com/opentable/kafka/builders/KafkaConsumerBuilder.java [hunk 5]
```diff
@@ -67,6 +67,27 @@ public KafkaConsumerBuilder(Map<String, Object> prop, EnvironmentProvider enviro
         kafkaBaseBuilder = new KafkaBaseBuilder(prop, environmentProvider);
     }
 
+    /**
+     * Cloning constructor. All properties must be copied here.
+     *
+     * @param src
+     */
+    private KafkaConsumerBuilder(KafkaConsumerBuilder src) {
+        this.kafkaBaseBuilder = src.kafkaBaseBuilder;
```
Comment (mikebell90) [inline src/main/java/com/opentable/kafka/builders/KafkaConsumerBuilder.java]: These things are prone to error and omission. This is a serious flaw.

File: src/main/java/com/opentable/kafka/builders/KafkaConsumerBuilder.java [hunk 6]
```diff
@@ -113,12 +139,12 @@ public KafkaConsumerBuilder(Map<String, Object> prop, EnvironmentProvider enviro
      * @param valDeSer value deserializer
      * @return this
      */
-    public KafkaConsumerBuilder<K, V> withDeserializers(Class<? extends Deserializer<K>> keyDeSer, Class<? extends Deserializer<V>> valDeSer) {
+    public <K2, V2> KafkaConsumerBuilder<K2, V2> withDeserializers(Class<? extends Deserializer<K2>> keyDeSer, Class<? extends Deserializer<V2>> valDeSer) {
```
Comment (mikebell90) [inline src/main/java/com/opentable/kafka/builders/KafkaConsumerBuilder.java]: Don't change this.

File: src/main/java/com/opentable/kafka/builders/KafkaConsumerBuilder.java [hunk 7]
```diff
@@ -113,12 +139,12 @@ public KafkaConsumerBuilder(Map<String, Object> prop, EnvironmentProvider enviro
...
         this.keyDe = keyDeSer;
         this.valueDe = valDeSer;
         this.keyDeserializerInstance = null;
         this.valueDeserializerInstance = null;
-        return this;
+        return new KafkaConsumerBuilder<>(this);
```
Comment (mikebell90) [inline src/main/java/com/opentable/kafka/builders/KafkaConsumerBuilder.java]: Nope, hate this.

File: src/main/java/com/opentable/kafka/builders/KafkaProducerBuilder.java [hunk 8]
```diff
@@ -46,10 +46,10 @@
     private OptionalLong lingerMS = OptionalLong.empty();
     private OptionalLong bufferMemory = OptionalLong.empty();
     private Class<? extends Partitioner> partitioner = DefaultPartitioner.class;
-    private Class<? extends Serializer<K>> keySe;
-    private Class<? extends Serializer<V>> valueSe;
-    private Serializer<K> keySerializer;
-    private Serializer<V> valueSerializer;
+    private Class<? extends Serializer> keySe;
+    private Class<? extends Serializer> valueSe;
+    private Serializer keySerializer;
```
Comment (mikebell90) [inline src/main/java/com/opentable/kafka/builders/KafkaProducerBuilder.java]: Yuck

**Comment by mikebell90** [comment]
If you really want to continue the discussion on the builders you may, but I don't see any improvement, and I see probable regression. Please revert that.

File: src/main/java/com/opentable/kafka/builders/KafkaBuilderConfiguration.java (hunk 2 above)
Comment (dkaukov) [reply src/main/java/com/opentable/kafka/builders/KafkaBuilderConfiguration.java]: done.

File: src/main/java/com/opentable/kafka/builders/KafkaConsumerBuilder.java (hunk 7 above)
Comment (dkaukov) [reply src/main/java/com/opentable/kafka/builders/KafkaConsumerBuilder.java]: I've removed cloning. If this was a concern. ;(

**Comment by dkaukov** [comment]
@mikebell90 Generally speaking, types are undefined until you specify Serializers/Deserializers. So I want to avoid:
```
        Consumer<Integer, String> res = builderFactoryBean.<Integer, String>builder("testme")
            .withDeserializers(IntegerDeserializer.class, StringDeserializer.class)
            .build();
```
1) You have to specify types 3 times
2) `builderFactoryBean.<Integer, String>builder("testme")` - syntax is really weird (atleast for me)
3) Auto-completion in IntelliJ prompting you to type `builderFactoryBean.builder("testme")` and than you are doomed for 5 mins.. ;)

I think, this is slightly better:
```
        Consumer<Integer, String> res = builderFactoryBean.consumerBuilder("testme")
            .withDeserializers(IntegerDeserializer.class, StringDeserializer.class)
            .build();
```
Feels more natural (for me after KafkaStreams, Spark, etc). Plus forces you to specify serializers/deserializers.

BTW: I've addressed  your concerns about cloning and type-safety.

File: src/main/java/com/opentable/kafka/builders/KafkaConsumerBuilder.java (hunk 6 above)
Comment (dkaukov) [reply src/main/java/com/opentable/kafka/builders/KafkaConsumerBuilder.java]: see comment

File: src/main/java/com/opentable/kafka/builders/KafkaBuilderFactoryBean.java (hunk 3 above)
Comment (dkaukov) [reply src/main/java/com/opentable/kafka/builders/KafkaBuilderFactoryBean.java]: see comment

**Pull request merged**